modelling/src/main/java/org/axonframework/modelling/SimpleEntityAsyncRepository.java

```diff
@@ -0,0 +1,181 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.modelling;
+
+import jakarta.annotation.Nonnull;
+import org.axonframework.common.infra.ComponentDescriptor;
+import org.axonframework.messaging.Context.ResourceKey;
+import org.axonframework.messaging.unitofwork.ProcessingContext;
+import org.axonframework.modelling.repository.AsyncRepository;
+import org.axonframework.modelling.repository.ManagedEntity;
+
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.UnaryOperator;
+
+/**
+ * Simple implementation of an {@link AsyncRepository} that can load and persist entities of a given type. When an
+ * entity is loaded, it is stored in the {@link ProcessingContext} to ensure that the same entity is returned when it is
+ * loaded again in the same context.
+ * <p>
+ * When an entity is successfully loaded, it's scheduled to automatically be persisted when the
+ * {@link ProcessingContext} is committed.
+ * <p>
+ * Both the {@link #load(Object, ProcessingContext)} and {@link #loadOrCreate(Object, ProcessingContext)} use the
+ * {@link SimpleEntityLoader} to load the entity. If you wish to create an entity when it does not exist, you can return
+ * a value when you have found none.
+ *
+ * @param <I> The type of the identifier of the entity.
+ * @param <T> The type of the entity.
+ * @author Mitchell Herrijgers
+ * @since 5.0.0
+ */
+public class SimpleEntityAsyncRepository<I, T> implements AsyncRepository.LifecycleManagement<I, T> {
+
+    private final ResourceKey<Map<I, CompletableFuture<ManagedEntity<I, T>>>> managedEntitiesKey =
+            ResourceKey.withLabel("SimpleEntityRepositoryManagedEntities");
+
+    private final Class<I> entityIdClass;
+    private final Class<T> entityClass;
+    private final SimpleEntityLoader<I, T> loader;
+    private final SimpleEntityPersister<I, T> persister;
+
+    /**
+     * Constructs a new simple {@link AsyncRepository} for entities of type {@code entityClass} with identifiers of type
+     * {@code entityIdClass}. The given {@code loader} is used to load entities and the given {@code persister} is used
+     * to persist entities.
+     *
+     * @param entityIdClass The type of the identifier of the entity.
+     * @param entityClass   The type of the entity.
+     * @param loader        The component capable of loading entities.
+     * @param persister     The component capable of persisting entities.
+     */
+    public SimpleEntityAsyncRepository(
```

Comment:
I think none of these fields are allowed to be `null`, right?

Reply:
Correct, adding!